File: packages/valkey-search-kit/src/encoding.ts
/**
 * Encode number[] as a little-endian Float32 Buffer.
 * Used to store embeddings as binary HSET field values.
 */
export function encodeFloat32(vec: number[]): Buffer {
  const buf = Buffer.alloc(vec.length * 4);
  for (let i = 0; i < vec.length; i++) {
    buf.writeFloatLE(vec[i], i * 4);
  }
  return buf;
}

File: packages/valkey-search-kit-py/betterdb_valkey_search_kit/encoding.py
from __future__ import annotations

import struct


def encode_float32(vec: list[float]) -> bytes:
    """Encode a float list as little-endian Float32 bytes.

    Used to store embeddings as binary HSET field values for KNN search.
    """
    return struct.pack(f"<{len(vec)}f", *vec)


def decode_float32(data: bytes) -> list[float]:
    """Decode little-endian Float32 bytes into a float list."""
    n = len(data) // 4
    return list(struct.unpack_from(f"<{n}f", data))
